json_schema_fuzz/schema_operations.py

```diff
@@ -1,45 +1,222 @@
-        elif schema_a[key] == schema_b[key]:
-            pass  # same leaf value
-        elif (
-                isinstance(schema_a[key], list)
-                and isinstance(schema_b[key], list)
-        ):
-            # Append lists together
-            schema_a[key].extend(schema_b[key])
-        else:
-            raise NotImplementedError(
-                f"Conflicting key {key} encountered in path {path}")
-    if in_place:
-        return None
+    """
+    Get list of values from dictionaries.
+    If it is not present in any dictionary, return the default value.
+    """
+    values = []
+    for dictionary in dictionaries:
+        value = dictionary.get(key, None)
+        if value is not None:
+            values.append(value)
+    if len(values) > 0:
+        return values
     else:
-        return schema_a
+        return default
+
+
+def get_index_or_default(
+        given_list: list,
+        index: int,
+        default=None
+):
+    """ Get index from a list or get a default value """
+    try:
+        return given_list[index]
+    except IndexError:
+        return default
+
+
+class SchemaIsFalse(Exception):
+    """ Raised if schema will not validate true for any values """
+
+
+def merge_listify(values):
+    """
+    Merge values by converting them to lists
+    and concatenating them.
+    """
+    output = []
+    for value in values:
+        if isinstance(value, list):
+            output.extend(value)
+        else:
+            output.append(value)
+    return output
+
+
+def merge_all_equal(values):
+    """
+    Merge values and raise SchemaIsFalse exception
+    if all values are not equal
+    """
+    if not values.count(values[0]) == len(values):
+        raise SchemaIsFalse()
+    return values[0]
+
+
+# pylint: disable=too-many-branches
+# pylint: disable=too-many-locals
+# pylint: disable=too-many-statements
+
+
+def merge(
+    *schemas: List[Dict[Any, Any]],
+) -> Dict[Any, Any]:
+    """
+    Merge a list of JSON schemas recursively.
+
+    Merging schemas here means that an instance
+    must validate against all provided schemas
+    in order to validate against the combined schema.
+
+    This is equivalent to an allOf with the provided schemas.
+    """
+
+    merged_schema = {}
+
+    # Dictionary of properties and how
+    # to merge them together
+    property_merging_functions = {
+        # Combinations
+        "allOf": merge_listify,
+
+        # Numbers
+        "maximum": min,
+        "exclusiveMaximum": min,
+        "minimum": max,
+        "exclusiveMinimum": max,
+        "multipleOf": lcm,
+        "notMultipleOf": merge_listify,
+
+        # String
+        "minLength": max,
+        "maxLength": min,
+
+        # Object
+        "required": merge_listify,
+        "anyAdditionalProperty": merge_listify,
+        "additionalProperties": lambda values: merge(*values),
+
+        # Array
+        "hasDuplicates": merge_all_equal,
+        "uniqueItems": merge_all_equal,
```

Comment:
I think this is wrong. See https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.4.3 and my other comment in the tests.

Reply:
If `uniqueItems: false` worked as this implies, I think we would not need `hasDuplicates`.

Reply:
I think we still need hasDuplicates based on the behavior of uniqueItems. The inverse of a schema is one that validates only on instances where the original schema is false. So if the original schema has uniqueItems, we still need to ensure that our inverse schema produces values with duplicates. I think we need to fix the merging behavior but I do think that we still need hasDuplicates.

Reply:
I agree.